vm/src/pyobject.rs

```diff
@@ -643,8 +643,13 @@ impl PyContext {
         objtype::new(self.type_type(), name, vec![base], PyAttributes::new()).unwrap()
     }
 
-    pub fn new_scope(&self) -> Scope {
-        Scope::new(None, self.new_dict())
+    pub fn new_scope(&self, vm: &VirtualMachine) -> Scope {
+        let scope = Scope::new(None, self.new_dict());
+
+        let globals = scope.globals.clone();
+        let _ = globals.set_item("__builtins__", vm.builtins.clone(), vm);
```

Comment:
I'm unsure whether this is the right place to set the item. Anyways, the function name should be changed, since the function now no longer creates a new scope. I would suggest to leave the original function in place, and create a second function which creates a scope (by calling the original function) and adds the builtin module. Then we can use this new function when we require a scope with builtins.

In other words: do we always require builtins in the scope?

Reply:
I think ```__builtins__``` get added in every case (except presumably during the creation of builtins itself). Even if we pass a dictionary to as globals, ```__builtins__``` gets added:

```
>>> exec('assert "__builtins__" in globals()', dict())
```

It doesn't get overwritten if already there:

```
exec('assert __builtins__ == 7', {'__builtins__': 7})
```

And the ```__builtins__``` provided is used for resolution of builtin names:

```
>>> exec('print()', {'__builtins__': {}})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 1, in <module>
NameError: name 'print' is not defined
```

And it doesn't use set_item to add it to the dictionary. (If you use a custom class for globals you don't get to intercept the addition of builtins.

There's also something odd that ```__builtins__``` is sometimes a module and sometimes a dict.

I'm not sure what the best implementation is. I lean towards not changing ```new_scope``` and doing the insert in the caller (because some callers will have to do special things to add it to globals that lack it.